File: Data Structures & Algorithms/validate-parentheses/submission-0.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        pairs = {
            '}': '{',
            ']': '[',
            ')': '(',
        }
        
        for c in s:
            if c in pairs.values():
                stack.append(c)
            elif c in pairs:
                if len(stack) == 0 or stack[-1] != pairs[c]:
                    return False
                stack.pop()

        return len(stack) == 0

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-0.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        int res = 0;
        int n = heights.size();
        stack<int> stk;

        for (int i = 0; i <= n; ++i){
            while (!stk.empty() && (i == n || heights[i] < heights[stk.top()])){
                int height = heights[stk.top()];
                stk.pop();

                int left = stk.empty() ? -1 : stk.top();
                int width = i - left - 1;

                res = max(res, height*width);
            }

            stk.push(i);
        }

        return res;
    }
};
